src/Elastic.Apm/Metrics/MetricsProvider/BreakdownMetricsProvider.cs

```diff
@@ -0,0 +1,87 @@
+﻿// Licensed to Elasticsearch B.V under
+// one or more agreements.
+// Elasticsearch B.V licenses this file to you under the Apache 2.0 License.
+// See the LICENSE file in the project root for more information
+
+using System.Collections.Generic;
+using System.Linq;
+using Elastic.Apm.Api;
+using Elastic.Apm.Helpers;
+using Elastic.Apm.Model;
+
+namespace Elastic.Apm.Metrics.MetricsProvider
+{
+	internal class BreakdownMetricsProvider : IMetricsProvider
+	{
+		internal const string SpanSelfTime = "span.self_time";
+
+		private readonly List<MetricSet> _itemsToSend = new();
+
+		private readonly object _lock = new();
+		private int _transactionCount;
+		public int ConsecutiveNumberOfFailedReads { get; set; }
+
+		public string DbgName => nameof(BreakdownMetricsProvider);
+
+		public bool IsMetricAlreadyCaptured
+		{
+			get
+			{
+				lock (_lock)
+					return _itemsToSend.Count > 0;
+			}
+		}
+
+		public bool IsEnabled(IReadOnlyList<WildcardMatcher> matchers) => !WildcardMatcher.IsAnyMatch(matchers, SpanSelfTime);
+
+		public void CaptureTransaction(Transaction transaction)
+		{
+			lock (_lock)
+			{
+				_transactionCount++;
+				var timestampNow = TimeUtils.TimestampNow();
+
+				foreach (var item in transaction.SpanTimings)
+				{
+					var metricSet =
+						new MetricSet(timestampNow,
+							new List<MetricSample>
+							{
+								new($"{SpanSelfTime}.count", item.Value.Count), new($"{SpanSelfTime}.sum.us", item.Value.TotalDuration * 1000)
+							})
+						{
+							Span = new SpanInfo { Type = item.Key.Type, SubType = item.Key.SubType },
+							Transaction = new TransactionInfo { Name = transaction.Name, Type = transaction.Type }
+						};
+
+					_itemsToSend.Add(metricSet);
+				}
+
+				var transactionMetric =
+					new MetricSet(timestampNow,
+						new List<MetricSample>
+						{
+							new("transaction.duration.count", _transactionCount),
+							new("transaction.duration.sum.us", transaction.Duration!.Value * 1000),
+							new("transaction.breakdown.count", _transactionCount),
+						}) { Transaction = new TransactionInfo { Name = transaction.Name, Type = transaction.Type } };
+
+				_itemsToSend.Add(transactionMetric);
+			}
+		}
+
+		public IEnumerable<MetricSet> GetSamples()
+		{
+			var retVal = new List<MetricSet>();
+
+			lock (_lock)
+			{
+				retVal.AddRange(_itemsToSend.Take(1000));
+				_itemsToSend.Clear();
```

Comment:
should `_itemsToSend` be cleared? What if it still contains items to send after taking 1000?

Reply:
> should _itemsToSend be cleared? What if it still contains items to send after taking 1000?

The spec says: "...agents must cap the total number of metricsets. A reasonable default seems to be  [1000]. When the limit is reached, agents should log sla warning...". So my understanding is that everything above 1k should be thrown away, so it's fine to clear those. Also worth noting that in this case we'd try to show time spent in percentage across 1K groups, so at this point I'd say the visualization can be also challenging.

On the other hand by thinking about this: we should not even collect those in `_itemsToSend`, plus the warning was also missing. So I moved the handling of this into `CaptureTransaction(Transaction)`: we only collect up to 1K items and if 1K is reached we log.

I also included a test to cover this.